internal/s3storage/s3storage.go

```diff
@@ -68,3 +163,60 @@ func (s *Service) CreateArtifact(artifactSpec artifact.Spec, md5 string) (string
 
 	return uploadURL, nil
 }
+
+func (s *Service) Close() error {
+	if s.sqsHandlerQuitChannel != nil {
+		close(s.sqsHandlerQuitChannel)
+	}
+	if s.sqsHandlerErrorChannel != nil {
+		err := <-s.sqsHandlerErrorChannel
+		return err
+	}
+	return nil
+}
+
+func (s *Service) startSQSHandler(handler func(msg string) error) {
+	log.Infof("starting SQS handler")
+	s.sqsHandlerQuitChannel = make(chan struct{})
+	s.sqsHandlerErrorChannel = make(chan error, 1)
+	go func() {
+		for {
+			_, ok := <-s.sqsHandlerQuitChannel
```

Comment:
The initial check on the `sqsHandlerQuitChannel` handler is only checked on startup. It should instead be in a select statement with `output.Messages` as to continually validate whether the service is being closed.

```golang
output, err := s.sqsClient.ReceiveMessage(&sqs.ReceiveMessageInput{
if err != nil {
	// handle
}
for {
	select {
		case <-s.sqsHandlerQuitChannel:
			return
		case message := <-output.Messages:
			// handle
	}
}
```

Reply:
Why is it only checked on startup? `s.sqsClient.ReceiveMessage(&sqs.ReceiveMessageInput` is executed every 5 seconds, which is then followed by `_, ok := <-s.sqsHandlerQuitChannel`.

Reply:
Rewrote this, so lets see if you still think its a problem :)